src/vcpkg/binarycaching.cpp

```diff
@@ -1020,34 +1020,243 @@ namespace
+                                .string_arg("BlobLocal")
+                                .string_arg("--output-level")
+                                .string_arg("QUIET")
+                                .string_arg("--overwrite")
+                                .string_arg("true")
+                                .string_arg(m_url.make_container_path())
+                                .string_arg(tmp_downloads_location)
+                                .string_arg("--include-path");
+
+            const size_t fixed_len = base_cmd.command_line().size() + 4; // for space + surrounding quotes + terminator
+            for (auto&& batch : batch_azcopy_args(abis, fixed_len))
+            {
+                auto maybe_output =
+                    cmd_execute_and_capture_output(Command{base_cmd}.string_arg(Strings::join(";", batch)));
+                // We don't return on a failure because the command may have
+                // only failed to restore some of the requested packages.
+                if (!maybe_output.has_value())
+                {
+                    msg::println_warning(maybe_output.error());
+                }
+            }
+
+            for (auto&& file : m_fs.get_files_recursive(tmp_downloads_location, VCPKG_LINE_INFO))
+            {
+                auto filename = file.stem().to_string();
+                auto it = abi_index_map.find(filename);
+                if (it != abi_index_map.end())
+                {
+                    out_zip_paths[it->second].emplace(std::move(file), RemoveWhen::always);
+                }
+            }
+        }
+
+        void precheck(View<const InstallPlanAction*> actions, Span<CacheAvailability> cache_status) const override
         {
+            auto abis = azcopy_list();
+            if (abis.empty())
+            {
+                // If the command failed, we assume that the cache is unavailable.
+                std::fill(cache_status.begin(), cache_status.end(), CacheAvailability::unavailable);
+                return;
+            }
+
+            for (size_t idx = 0; idx < actions.size(); ++idx)
+            {
+                auto&& action = *actions[idx];
+                const auto& abi = action.package_abi().value_or_exit(VCPKG_LINE_INFO);
+                cache_status[idx] =
+                    Util::contains(abis, abi) ? CacheAvailability::available : CacheAvailability::unavailable;
+            }
         }
 
-        std::string make_object_path(const std::string& prefix, const std::string& abi) const override
+        LocalizedString restored_message(size_t count,
+                                         std::chrono::high_resolution_clock::duration elapsed) const override
         {
-            return Strings::replace_all(prefix, "{sha}", abi);
+            return msg::format(
+                msgRestoredPackagesFromAzureStorage, msg::count = count, msg::elapsed = ElapsedTime(elapsed));
         }
+
+        Path m_buildtrees;
```

Comment:
No change requested here because HttpGetBinaryProvider already does this, but it seems kinda strange to me that we use buildtrees for the temporary zip on 'get' but packages for the temporary zip on 'push'. It seems like we should use packages for both?